### ISSUE TITLE
Zero downtime deployment of worker groups

### ISSUE THREAD

**Issue by geota** [terraform-aws-modules/terraform-aws-eks#185]
# I have issues performing a zero downtime deployment of my worker group ASGs because of how this module couples/manages the worker group asgs resources.

## I'm submitting a...

* [ ] bug report
* [X] feature request
* [ ] support request
* [ ] kudos, thank you, warm fuzzy

## What is the current behavior?
Currently, making changes to your ASGs can cause the ASGs to get recreated. The new ASGs are created BEFORE the existing ones are destroyed BUT there is not ability to pause the destruction while pods are drained from the old ASGs worker nodes to the newly deployed ASGs worker nodes.

This is largely because the worker group ASGs are managed inline with this module and are managed as a list of maps. If any of the indices change for an existing worker group config TF is not smart enough to know that and attempts to recreate them.

For example - if I had a 3 worker group defined. Then I want to update ALL of the existing worker groups. I could add 3 new ones, update the count to 6 and TF would provision the three new ASGs and NOT touch the pre-existing three. I can then drain my old nodes onto my new nodes. However, if I then prune the list of the old worker groups and drop the count back to 3 - TF trys to RECREATE my existing ASGs and tear down the newly created ones because indices 4,5,6 have disappeared and the contents of 1,2,3 have changed.

## What's the expected behavior?
Ideally I would like an easy way to provision new ASGs and KEEP my old ones around until I explicitly decide to delete them (after I have drained the old nodes and verified my services are up).

My current work around is to use ```terraform mv``` to move the worker resources (instance profiles, launch configurations, and asgs) state into a new module and new state file so that terraform doesnt know about the pre-existing workers.

I then make my changes to my Terraform to add the new set of worker groups and run terraform apply. I then can drain my old nodes onto my newly created ASGs. After I have verified everything is healthy, I then can create a dummy terraform main.tf and point it at the new state file I moved the pre-existing worker group resource state into and let terraform tear it down.

This is a bit hacky and it seems this module should provide someway of doing zero downtime upgrades of your worker groups.

## Environment details

* Affected module version:
* OS:
* Terraform version:

**Comment by mmcaya** [comment]
Does the comment from https://github.com/terraform-aws-modules/terraform-aws-eks/issues/82#issuecomment-411670234 give you a reasonable path forward?

TL;DR Reuse you worker groups in a blue / green fashion, having a deterministic pattern for worker node change by altering the ASG counts after draining and rebalancing.

**Comment by max-rocket-internet** [comment]
> This is largely because the worker group ASGs are managed inline with this module and are managed as a list of maps. If any of the indices change for an existing worker group config TF is not smart enough to know that and attempts to recreate them.

I get you, it's massive pain, but I think this is general terraform problem though. One that has existed for a very long time.

A simple example:

```
variable "envs" {
  default = [
    "stg01",
    "stg02",
    "stg03",
  ]
}

resource "aws_sns_topic" "test" {
  count = "${length(var.envs)}"
  name  = "test-${element(var.envs, count.index)}"
}
```

If you remove the `stg02` item from the list and run `terraform apply` then 2 resources are removed and 1 is recreated.

@mmcaya's workaround is where basically you are never removing any resources, you are just scaling down the ASGs.

You mention draining a lot but this process only exists inside k8s, not in Terraform, so not sure what else we can do in this module? I'm open to suggestions though!

**Comment by geota** [comment]
@max-rocket-internet  it sorta works but still is very clunky.

The workflow I landed on....

Explicitly move the terraform state into another module/state file for the ASGs, launch configs, etc so terraform forgets about these assets.
Apply the terraform update to create your new set of ASGs.
Explicitly drain the old instances/nodes.
Perform a terraform teardown and point it at the state file you moved the original ASGs/etc assets into.

This works well enough- feel free to close this issue. Its probably not worth trying to solve this until terraform 0.12 is released and we can better primitives to build off of.
